feat(api, robot-server): get historical and current command errors

TamarZanzouri opened the pull request:
# Overview

closes https://opentrons.atlassian.net/browse/EXEC-655.
GET historical command errors and refactor current command errors.

## Test Plan and Hands on Testing

run a protocol with ER failed commands.
- GET `/runs/{runId}/commandsErros` while the run is active and make sure you get the failed commands errors.
- GET `/runs/{runId}/commandsErros` when the run is finished and make sure you get the failed commands errors.

## Changelog

- added db schema 8.
- added `command_error`, `command_status` field to `commands_table`.
- insert `command_error`, `command_status` when inserting a command to commands table.
- `get_commands_errors_slice`
- added missing indexes from previous migrations. 

## Review requests

changes make sense?

## Risk assessment

low.

TamarZanzouri committed:
WIP intial schema and orchestrator changes

TamarZanzouri committed:
changed schema and imports to 8

TamarZanzouri committed:
insert the errors to the DB

TamarZanzouri committed:
refactor into commands table

TamarZanzouri committed:
select from commands

TamarZanzouri committed:
linting and command error slice done

TamarZanzouri committed:
added tests and fixed bugs

TamarZanzouri committed:
changed logic for current command errors

TamarZanzouri committed:
removed failed_command_errors from state

TamarZanzouri committed:
refactor get command errors

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

TamarZanzouri commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -127,6 +133,7 @@ def update_run_state(
             run_id: The run to update
             summary: The run's equipment and status summary.
             commands: The run's commands.
+            command_errors:  The run's commands errors.
```
```suggestion
```

TamarZanzouri committed:
migration and initialization fixes

TamarZanzouri committed:
Update robot-server/robot_server/runs/run_store.py

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

TamarZanzouri committed:
change get_run_errors to get count

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

TamarZanzouri committed:
tables tests fixed

TamarZanzouri committed:
added test for historical run

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

sanni-t commented on robot-server/robot_server/persistence/persistence_directory.py:
```diff
@@ -55,7 +55,8 @@ async def prepare_active_subdirectory(prepared_root: Path) -> Path:
             # schema that was never released to the public. It may be present on
             # internal robots.
-            v6_to_v7.Migration6to7(subdirectory=LATEST_VERSION_DIRECTORY),
+            v6_to_v7.Migration6to7(subdirectory="7"),
```
This should be 7.1 as that seems like the last directory that was used under schema 7

sanni-t commented on robot-server/robot_server/persistence/file_and_directory_names.py:
```diff
@@ -8,7 +8,7 @@
 from typing import Final
 
-LATEST_VERSION_DIRECTORY: Final = "7.1"
+LATEST_VERSION_DIRECTORY: Final = "8.1"
```
This should be just 8.

TamarZanzouri committed:
fixed persistance path

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/command_history.py:
```diff
@@ -251,6 +262,10 @@ def _add(self, command_id: str, command_entry: CommandEntry) -> None:
+    def append_failed_command_id(self, command_id: str) -> None:
+        """Create or update a failed command id."""
+        self._all_failed_command_ids.append(command_id)
+
```
Let's not make this a public method, if possible. My thinking was that `CommandHistory` could do this automatically inside `set_command_failed`.

The fact that methods like this exist for the other fields is a historical thing, left over from quickly refactoring, that I think we should try to move away from.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,86 @@
+
+- Adds a new command_intent to store the commands intent in the commands table
+- Adds a new source to store the data files origin in the data_files table
+- Adds the `boolean_setting` table.
```
Reminder to update this docstring.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,86 @@
+
+        dest_db_file = dest_dir / DB_FILE
+
+        # Append the new column to existing protocols and data_files in v6 database
```
Outdated comment.

SyntaxColoring commented on robot-server/tests/persistence/test_tables.py:
```diff
@@ -123,6 +124,9 @@
     CREATE INDEX ix_protocol_protocol_kind ON protocol (protocol_kind)
     """,
     """
+    CREATE INDEX ix_run_command_command_error ON run_command (command_error)
+    """,
```
I have not tested this, but I don't think this index is what we want. (This index is what happens when you just do `index=True` on the new column declaration.)

This will, across all commands from all runs, maintain an index that is a lexicographic sort of the errors' JSON strings.

That does not help with efficiently serving the `SELECT` statement that you've written in `RunStore`, which is something like, "get all the errors from a single run, in order."

So I think what we want instead is either:

* A compound index on `(run_id, has_error, index_in_run)`, where `has_error` is a true/false value computed from `command_error` being non-`NULL`. I don't know how to do this off the top of my head with SQLAlchemy, but [here is SQLite documentation that looks relevant](https://www.sqlite.org/expridx.html).
* A compound index on `(run_id, command_status, index_in_run)`, where `command_status` is a new column storing the command's `succeeded`/`failed` status. And then the `SELECT` statement filters based on `command_status == "failed"`.

SyntaxColoring replied:
We could also test whether it's fast enough to just use the existing `(run_id, index_in_run)` compound index. From there, filtering for just the failed commands will be an O(n) linear scan, but...it's a linear scan implemented by optimized C, so maybe it's good enough.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,86 @@
+
+            dest_transaction = exit_stack.enter_context(dest_engine.begin())
+
+            def add_column(
+                engine: sqlalchemy.engine.Engine,
+                table_name: str,
+                column: Any,
+            ) -> None:
+                column_type = column.type.compile(engine.dialect)
+                engine.execute(
+                    f"ALTER TABLE {table_name} ADD COLUMN {column.key} {column_type}"
+                )
+
+            add_column(
+                dest_engine,
+                schema_8.run_command_table.name,
+                schema_8.run_command_table.c.command_error,
+            )
```
~~Because your new column is nullable, I don't think you need any of this `add_column()` stuff. `metadata.create_all()` can add the column to the existing table in-place, as long as the column is nullable.~~

TamarZanzouri replied:
@SyntaxColoring when removing this all the tests are failing on a `command_error` column does not exist so I think we do need this.

SyntaxColoring replied:
I stand corrected. I guess it adds missing tables but not missing columns.

In that case, can we deduplicate this `add_column()` helper, which is copy-pasted across 4 migrations, by moving it to `_util.py`?

SyntaxColoring replied:
Wait, `metadata.create_all()` isn't adding new columns, but ~~it *is* adding new indexes~~? [edit: no it's not, and that's another thing we need to figure out.] That seems weird, no?

TamarZanzouri replied:
that does seem weird, I will look into this.

TamarZanzouri replied:
@SyntaxColoring I looked into this, if we do a `metadata.drop_all()` before we `create_all` it works. apparently it only creates new tables and not added columns. and bc we do not want to drop all and lose our data we must use the `add_column` logic.

SyntaxColoring replied:
I agree that it doesn't make sense to use `.drop_all()`.

But I don't think this existing pattern is working, either. Have you seen the schema that this is outputting? Like, try running `sqlite3 robot_server.db .schema` on a `robot_server.db` file that was created by this migration path. It will show something like this:

```sql
/* ... */
CREATE TABLE run_command (
	row_id INTEGER NOT NULL,
	run_id VARCHAR NOT NULL,
	index_in_run INTEGER NOT NULL,
	command_id VARCHAR NOT NULL,
	command VARCHAR NOT NULL, command_intent VARCHAR, command_error VARCHAR, command_status VARCHAR(9),
	PRIMARY KEY (row_id),
	FOREIGN KEY(run_id) REFERENCES run (id)
);
/* ... */
```

The new `command_error` column is there, because you're manually creating it with the custom `add_column()` thing. But the new constraint `commandstatussqlenum` and the new index `ix_run_run_id_index_in_run_command_status` are *not* there, because `metadata.create_all()` skips the preexisting table, I guess.

`test_tables.py` isn't catching this bug because it assumes we create new database file with `metadata.create_all()`. That was true for some of the previous migrations, but it's not true for this one.

SyntaxColoring replied:
So I see a few options:

* Create the new file from scratch, then open it as a second SQLAlchemy connection, then create the schema inside it with `metadata.create_all()`, and then copy all the data over explicitly.

  Some of the earlier migrations work like this. It's slower because all the data needs to pass through Python. That may or may not matter—we'd need to test it.

* Create the new file from scratch, then open it in the *original* SQLAlchemy connection with [`ATTACH DATABASE`](https://www.sqlite.org/lang_attach.html), then create the schema inside it with `metadata.create_all()`, and then copy all the data over explicitly.

  This is like the above, but it should be faster because the data never has to enter Python. I looked into this a while ago but couldn't figure out how to get the `ATTACH DATABASE` thing working without basically hand-writing every SQL query—SQLAlchemy doesn't seem to support this quirk of SQLite very well.

* Create the new file by copying the old file, then rename the old tables, then create the new table, indexes, and constraints with `metadata.create_all()`, then copy the data over explicitly.

  This is basically equivalent to the above, except we only have to do weird stuff with one table instead of the whole database, so maybe it's easier. I haven't tried.

* Add something like `add_column()` that works on constraints and indexes and anything else that `metadata.create_all()` is skipping.

Regardless of which we choose, I think we really need to fix `test_tables.py` to ensure parity between what's written in `schema_8.py` and what the migration path actually outputs.

SyntaxColoring replied:
Recap from call:

* We'll do this:

   > Add something like add_column() that works on constraints and indexes and anything else that metadata.create_all() is skipping.

* We can probably remove the `metadata.create_all()`—it's not doing anything and it might be misleading

* And in a separate PR, I'll try to get the tests to cover this. EXEC-827

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,86 @@
+
+        dest_transaction.execute(
+            f"UPDATE run_command SET command_error='{new_command_error}' WHERE row_id={row.row_id}"
+        )
```
Recap from discussion in-person:

Let's not manually construct raw SQL strings like this. We should be able to do this with `sqlalchemy.update()`. This will fail with a SQL syntax error if `new_command_error` contains any special characters, which the JSON strings certainly will.

Our migrations have sometimes needed to use raw SQL strings for things that SQLAlchemy does not support, like `ALTER TABLE` statements. That's tolerable because the input is trusted and hard-coded, unlike here.

SyntaxColoring reviewed (changes requested):
Thank you!

TamarZanzouri committed:
fixed import from schema 7

TamarZanzouri committed:
rollback add_column change

TamarZanzouri committed:
append_failed_command_id when setting a failed command

TamarZanzouri committed:
append_failed_command_id when setting a failed command

TamarZanzouri committed:
moved test to command state

TamarZanzouri committed:
changed index and lint

TamarZanzouri committed:
WIP - command status

TamarZanzouri committed:
fixed failing tests

TamarZanzouri committed:
insert and get with command_status

TamarZanzouri committed:
translate to CommandStatusSQLEnum

TamarZanzouri committed:
nips

SyntaxColoring commented on robot-server/robot_server/runs/router/base_router.py:
```diff
@@ -526,14 +526,13 @@ async def get_run_commands_error(
     try:
-        all_errors = run_data_manager.get_command_errors(run_id=runId)
-        total_length = len(all_errors)
+        all_errors_count = run_data_manager.get_command_errors_count(run_id=runId)
```
This will work for now and I don't think we should change it here, but thinking ahead, we should aim to read `all_errors_count` and `command_error_slice` in the same transaction.

This didn't matter before, when were were just reading stuff from memory. But now we're hitting the filesystem, which we'll probably want to make `async` at some point, so we should be careful about patterns that will open us up to torn-read concurrency hazards. Plus the small overhead of opening a transaction.

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_command_state.py:
```diff
@@ -1100,3 +1101,99 @@ def test_get_state_update_for_false_positive() -> None:
+
+
+def test_get_errors_slice_empty() -> None:
+    """It should return a slice from the tail if no current command."""
```
Does this docstring need updating?

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_command_state.py:
```diff
@@ -1100,3 +1101,99 @@ def test_get_state_update_for_false_positive() -> None:
+    # error_1 = ErrorOccurrence.construct(id="error-id-1")  # type: ignore[call-arg]
+    # error_2 = ErrorOccurrence.construct(id="error-id-2")  # type: ignore[call-arg]
+    # error_3 = ErrorOccurrence.construct(id="error-id-3")  # type: ignore[call-arg]
+    # error_4 = ErrorOccurrence.construct(id="error-id-4")  # type: ignore[call-arg]
```
Leftover code from debugging?

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -179,6 +186,10 @@ def update_run_state(
+                        "command_error": pydantic_to_json(command.error)
+                        if command.error
+                        else None,
+                        "command_status": CommandStatusSQLEnum(command.status.value),
```
This works, but as a general pattern, when we map between enums like this, I think we should usually do it with `if`/`else` ladders or `match` statements, for type safety. Constructing one enum from the other one's `.value` basically ties them together, negating the benefits of splitting them up.

For example, if we added a new command status like `cancelled`, this would silently break. Mypy isn't smart enough to notice that `"cancelled"` will raise if you pass it to `CommandStatusSQLEnum()`.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,96 @@
+) -> None:
+    """Add a new 'command_intent' column to run_command_table table."""
+    commands_table = schema_8.run_command_table
+    select_commands = sqlalchemy.select(commands_table).order_by(sqlite_rowid)
```
Nit: I'm pretty sure we can remove this `.order_by(sqlite_rowid)` here. The commands table already has a proper column for ordering, `index_in_run`.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,96 @@
+    commands_table = schema_8.run_command_table
+    select_commands = sqlalchemy.select(commands_table).order_by(sqlite_rowid)
+
+    for row in dest_transaction.execute(select_commands).all():
```
This will send one `UPDATE` statement for every command. So, potentially, tens of thousands of `UPDATE` statements. How bad is that time overhead going to be? Do we need to worry about batching this?

If there's a major performance issue, we don't need to fix it in this PR, but we would need to ticket it for the next release.

TamarZanzouri replied:
will look into batch update

TamarZanzouri replied:
https://stackoverflow.com/questions/25694234/bulk-update-in-sqlalchemy-core-using-where

SyntaxColoring replied:
I'm glad we figured out a way to do this, but with a batch update, I'd worry about the memory footprint.
(When I said "how bad is that time overhead going to be?" and "do we need to worry about batching this?", I really was neutrally asking those questions, not necessarily saying we should actually do a batch update.)

[Previously](https://github.com/Opentrons/opentrons/blob/383936698d9b90dab0b218cba72a04436e1939e0/robot-server/robot_server/persistence/_migrations/up_to_3.py#L204-L210), we've justified this kind of thing with napkin math estimating the memory footprint. Can we do the same here?

SyntaxColoring commented on robot-server/tests/runs/test_run_store.py:
```diff
@@ -289,6 +327,46 @@ async def test_update_run_state(
+        length=len(protocol_commands_errors),
+        cursor=0,
+    )
+    print(command_errors_result)
```
Left over from debugging?

```suggestion
```

SyntaxColoring commented on robot-server/tests/runs/test_run_store.py:
```diff
@@ -99,6 +100,43 @@ def protocol_commands() -> List[pe_commands.Command]:
 
+@pytest.fixture
+def protocol_commands_errors() -> List[pe_commands.Command]:
+    """Get a StateSummary value object."""
```
Outdated docstring?

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_8.py:
```diff
@@ -0,0 +1,338 @@
+        "index_in_run",
+        unique=True,
+    ),
+    # TODO(tz, 12-11-2024): change index to command_error when partial index is supported.
```
It turns out that SQLAlchemy does support partial indexes for SQLite. [Here's the relevant documentation.](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#partial-indexes)

I think I prefer the index on `command_status`, though. It seems more generally useful, and it matches how `CommandHistory` works. But I'm open to doing it the partial index way, too.

TamarZanzouri replied:
I was looking for this and could only find `postgres_where`! I am good with leaving as is

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_8.py:
```diff
@@ -0,0 +1,338 @@
+        unique=True,
+    ),
+    # TODO(tz, 12-11-2024): change index to command_error when partial index is supported.
+    sqlalchemy.Index(
+        "ix_run_run_id_index_in_run_command_status",  # An arbitrary name for the index.
+        "run_id",
+        "index_in_run",
+        "command_status",
+    ),
```
```suggestion
    sqlalchemy.Index(
        "ix_run_run_id_index_in_run_command_status",  # An arbitrary name for the index.
        "run_id",
        "index_in_run",
        "command_status",
    ),
```

1. This can be `unique=True`.
2. I don't think the order `(run_id, index_in_run, command_status)` is right, is it? This will sort first by `run_id`, then by `index_in_run`, then by `command_status`—and since no two commands with different `command_status` values will ever share the same `index_in_run` value, this will be equivalent to `(run_id, index_in_run)`. I think we want `(run_id, command_status, index_in_run)` instead—sort first by the run ID, then by the command status, and finally by the index in the run.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,96 @@
+"""Migrate the persistence directory from schema 7 to 8.
+
+Summary of changes from schema 7:
+
+- Adds a new command_error to store the commands error in the commands table
```
Reminder to also update this changelog for the `command_status` column.

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_8.py:
Since this is an intentional change to the server's behavior, we should probably make the commit title `feat` or `fix` instead of `chore`.

SyntaxColoring reviewed (commented):
Looks good, just the ongoing discussion above, a thing about the ordering of columns in the compound index, and some smaller things here.

TamarZanzouri committed:
Update robot-server/tests/runs/test_run_store.py

Co-authored-by: Max Marrone <max@opentrons.com>

TamarZanzouri committed:
nits

TamarZanzouri committed:
batch edit

TamarZanzouri committed:
batch update without binding

TamarZanzouri committed:
test table with new index and migration bulk fix

TamarZanzouri committed:
conditional update

TamarZanzouri committed:
map command status to sql command status

TamarZanzouri committed:
map command status to sql command status in migration

TamarZanzouri committed:
map command status to sql command status in migration

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -812,3 +814,19 @@ def _convert_state_to_sql_values(
     }
+
+
+def _convert_commands_status_to_sql_command_status(
+    status: str,
+) -> CommandStatusSQLEnum:
+    match status:
+        case "queued":
+            return CommandStatusSQLEnum.QUEUED
+        case "running":
+            return CommandStatusSQLEnum.RUNNING
+        case "failed":
+            return CommandStatusSQLEnum.FAILED
+        case "succeeded":
+            return CommandStatusSQLEnum.SUCCEEDED
+        case _:
+            assert False, "command status is unknown"
```
You'll want this `status` arg to be an enum, not a `str`, and then you'll want to remove the `case _` default, so the type checker will error if you forget to handle one of the input cases. That's what makes it type-safe. Otherwise, this would have the same disadvantages as doing `EnumA(enum_b.value)`.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -85,12 +86,42 @@ def _migrate_command_table_with_new_command_error_col_and_command_status(
+        )
+        commands_to_update.append(
+            {
+                "_id": row.row_id,
```
Nitpick: Does `"_id"` really need to have that underscore? Or is it only there because that's how the column was named in the example code?

TamarZanzouri replied:
i think that as long as we dont use a defined prop name we are good

TamarZanzouri committed:
add missing indexes

TamarZanzouri committed:
lint

TamarZanzouri committed:
change conversion into enum

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

TamarZanzouri committed:
fixing merge conflicts - WIP

TamarZanzouri committed:
WIP - fix test_tables failing

SyntaxColoring committed:
Fix up test_tables.py after merge.

Ensure EXPECTED_STATEMENTS_V8 is based on a copy-paste of the latest EXPECTED_STATEMENTS_V7.

SyntaxColoring committed:
Fix schema_8.py after merge.

Ensure it's based on a copy-paste of the latest (post-merge) schema_7.py.

SyntaxColoring commented:
This is pending some merge conflict resolution with `edge`. @TamarZanzouri and I will pick this up again when 8.2.0 release stuff chills out. Re-drafting until then.

SyntaxColoring committed:
Remove attempts to fix prior migrations, for now.

TamarZanzouri committed:
update test table

SyntaxColoring committed:
Make command_status column nullable to match what the migration does.

SyntaxColoring committed:
Remove enum constraint, for now.

TamarZanzouri committed:
Revert "update test table"

This reverts commit 3a29bfdb3d3a8ce46e719aaf0076c6148c5be6e0.

SyntaxColoring committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

SyntaxColoring commented:
Notable changes that I made when resolving merge conflicts:

1. Made `command_status` nullable to match the `ALTER TABLE` statement that you were doing in the migration. See my review comments below—we might want to treat that as a thing to fix, or we might not.
2. Removed the `ALTER TABLE` statements that you added to make up for the prior buggy migrations (EXEC-827). I think that was the right place to add them, but it was just feeling too much to have in one PR. Let's do it in a follow-up.
3. Removed the `command_status` enum constraint. The migration was not adding it, and when I tried to add it to the migration, I got confusing errors in `test_tables.py`. I think it's fixable, but I ran out of steam. Let's try it in a follow-up.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,135 @@
+                    f"ALTER TABLE {table_name} ADD COLUMN {column.key} {column_type}"
+                )
+
+            add_column(
+                dest_engine,
+                schema_8.run_command_table.name,
+                schema_8.run_command_table.c.command_error,
+            )
+
+            add_column(
+                dest_engine,
+                schema_8.run_command_table.name,
+                schema_8.run_command_table.c.command_status,
+            )
+
+            _add_missing_indexes(dest_transaction=dest_transaction)
+
+            _migrate_command_table_with_new_command_error_col_and_command_status(
+                dest_transaction=dest_transaction
+            )
```
No action needed, just some thoughts.

Early on in this PR, we made the decision to add the new column with `ALTER TABLE`, instead of letting SQLAlchemy create the new one from scratch.

I don't think that decision is panning out well.

1. We have to manually create indexes.
2. We have to manually create constraints. (And we're not doing that right now—see my `# todo` comment in `schema_8.py`).
3. The `command_status` column needs to be nullable even when that's not a good match for the data we're storing in it.

1 and 2 are thankfully caught by tests now, and perhaps we can improve them by fleshing out the `add_column()` function. But 3 is fundamental to SQLite. We talked about fixing 3 with Alembic, but I've looked into it and I'm skeptical that that will be easy.

Meanwhile, is `ALTER TABLE` really saving that much overhead compared to creating the new table from scratch? We're still iterating over every single command and parsing it as JSON, and I imagine that cost will dominate the savings from not having to copy the data.

I think we should stick with what you have now, but in the future, let's try the other way?

TamarZanzouri replied:
the other way means creating all tables from scratch? we still need to insert the data.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,135 @@
+        new_command_error = (
+            # Account for old_row.command["error"] being NULL.
+            None
+            if "error" not in row.command or data["error"] == None  # noqa: E711
```
Nitpick: Can we do `is None` here and avoid the `noqa: E711`, or is there specific reason to avoid it?

```suggestion
            if "error" not in row.command or data["error"] is None
```

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,135 @@
+    for row in dest_transaction.execute(select_commands).all():
+        data = json.loads(row.command)
+        new_command_error = (
+            # Account for old_row.command["error"] being NULL.
```
Nitpick: I think this comment is a little bit misleading. You're accounting for the *JSON field* `error` to be `null`, the lowercase JSON value, not `NULL`, the uppercase SQL value.

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -38,6 +45,7 @@
 
 from .action_models import RunAction, RunActionType
 from .run_models import RunNotFoundError
+from ..persistence.tables.schema_8 import CommandStatusSQLEnum
```
Nitpick: Can we import this from `persistence.tables` instead of `persistence.tables.schema_8` (re-export it through the `__init__.py`)? Otherwise this will be easy to forget when we have a schema 9.

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -537,6 +551,95 @@ def get_all_commands_as_preserialized_list(
+            count_result: int = transaction.execute(select_count).scalar_one()
+
+            actual_cursor = cursor if cursor is not None else count_result - length
+            # Clamp to [0, count_result).
```
This is beyond the scope of this PR, but I feel like this clamping and pagination logic has appeared in a few places across `api` and `robot-server`. Can we think about how to centralize it?

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -537,6 +551,95 @@ def get_all_commands_as_preserialized_list(
+                sqlalchemy.select(
+                    run_command_table.c.index_in_run,
+                    run_command_table.c.command_error,
+                )
```
I think you can do this:

```suggestion
                sqlalchemy.select(
                    run_command_table.c.command_error,
                )
```

Since you're not returning `index_in_run`.

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -537,6 +551,95 @@ def get_all_commands_as_preserialized_list(
+                )
+                .where(
+                    and_(
+                        run_command_table.c.run_id == run_id,
+                        run_command_table.c.index_in_run >= actual_cursor,
+                        run_command_table.c.index_in_run < actual_cursor + length,
+                        run_command_table.c.command_error is not None,
+                    )
```
I don't think this query is quite right? If someone's asking for errors starting from the 5th, this will return errors starting from the 5th *overall command,* but we probably want it to start from the 5th *failed command.*

Same for the page length. Non-failed commands will count against the page length, so if the client asks for 5 errors, the server will almost never return that many.

One way to do this would be with `LIMIT` and `OFFSET`. I am curious whether the query planner will handle those efficiently here.

(We should try to cover this in the `RunStore` unit tests. They're not catching this now because they're only testing a sequence of *just* failed commands—they'd need to test a sequence that has a mixture of failed and non-failed.)

SyntaxColoring replied:
Although, I see now that your docstring *does* say "the index in the whole collection." Is that intentional?

TamarZanzouri replied:
good catch! no that is not intentional! I will fix this with a sub select

SyntaxColoring reviewed (changes requested):
Mostly looks good, but I'm not sure about the query.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v7_to_v8.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def _add_missing_indexes(dest_transaction: sqlalchemy.engine.Connection) -> None:
+    # todo(2024-11-20): Probably add the indexes missing from prior migrations here.
+    # https://opentrons.atlassian.net/browse/EXEC-827
```
@TamarZanzouri FYI I think that it was correct for you to have had the other fixes in here, but when I was resolving merge conflicts, it was feeling like a bit too much to try to do at once, so I removed them. Let's do it in a follow-up?

SyntaxColoring commented on robot-server/robot_server/persistence/tables/schema_8.py:
```diff
@@ -0,0 +1,358 @@
+            CommandStatusSQLEnum,
+            values_callable=lambda obj: [e.value for e in obj],
+            validate_strings=True,
+            # nullable=True because it was easier for the migration to add the column
+            # this way. This is not intended to ever be null in practice.
+            nullable=True,
+            # todo(mm, 2024-11-20): We want create_constraint=True here. Something
+            # about the way we compare SQL in test_tables.py is making that difficult--
+            # even when we correctly add the constraint in the migration, the SQL
+            # doesn't compare equal to what create_constraint=True here would emit.
+            create_constraint=False,
```
@TamarZanzouri FYI, I made these changes when resolving merge conflicts:

1. Made `command_status` nullable to match the `ALTER TABLE` statement that you were doing in the migration.
2. Removed the `command_status` enum constraint because I couldn't figure out how to get the migration to match it in a way that would pass the tests. We should figure that out.

TamarZanzouri replied:
tried setting `create_constraint=True` but its acting weird in the test as you mentioned. maybe we should add a CREATE CONSTRAINT to the migration script?

TamarZanzouri committed:
Update robot-server/robot_server/persistence/_migrations/v7_to_v8.py

Co-authored-by: Max Marrone <max@opentrons.com>

TamarZanzouri committed:
Update robot-server/robot_server/runs/run_store.py

Co-authored-by: Max Marrone <max@opentrons.com>

TamarZanzouri committed:
select subquery and minor fixes

TamarZanzouri commented:
tested the migration on a FLEX and it works as expected. @SyntaxColoring

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -613,19 +613,26 @@ def get_commands_errors_slice(
             actual_cursor = cursor if cursor is not None else count_result - length
             # Clamp to [0, count_result).
             actual_cursor = max(0, min(actual_cursor, count_result - 1))
+            select_command_errors = (
+                sqlalchemy.select(run_command_table)
+                .where(run_command_table.c.command_error is not None)
+                .subquery()
+            )
```
Oh, heads up, this is a SQLAlchemy trap. Python doesn't let SQLAlchemy overload the `is not` operator, so we have to use [is_not()](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.ColumnElement.is_not).

Right now, this is equivalent to `.where(True)`. This should be easy to catch in `test_update_run_state_command_with_errors()` unit tests if you add some input commands that aren't `status: "failed"`.

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -537,6 +551,102 @@ def get_all_commands_as_preserialized_list(
+                .where(run_command_table.c.command_error is not None)
+                .subquery()
+            )
+            select_slice = (
+                sqlalchemy.select(run_command_table.c.command_error)
+                .where(
+                    and_(
+                        run_command_table.c.run_id == run_id,
+                        run_command_table.c.index_in_run >= actual_cursor,
+                        run_command_table.c.index_in_run < actual_cursor + length,
+                    )
+                )
+                .join_from(
+                    run_command_table,
+                    select_command_errors,
+                    onclause=run_command_table.c.index_in_run
+                    == select_command_errors.c.index_in_run,
+                )
+                .order_by(run_command_table.c.index_in_run)
+            )
+            slice_result = transaction.execute(select_slice).all()
```
Joins are not my strong suit, but is this right? It looks like we're still doing `run_command_table.c.index_in_run >= actual_cursor`. Don't we want `actual_cursor` to decide the offset within `select_command_errors`?

TamarZanzouri committed:
fixed join logic and added logic to the tests

TamarZanzouri committed:
fixed logic of cursor

TamarZanzouri committed:
remove join

SyntaxColoring commented on robot-server/robot_server/runs/run_store.py:
```diff
@@ -537,6 +551,106 @@ def get_all_commands_as_preserialized_list(
+                        == CommandStatusSQLEnum.FAILED,
+                    )
+                )
+                .subquery()
```
I think this subquery needs an `ORDER BY` clause too, right? Because you're depending on the incrementing numbers from `sqlalchemy.func.row_number()` corresponding to the actual order of the commands.

```suggestion
                .order_by(run_command_table.c.index_in_run)
                .subquery()
```

SyntaxColoring reviewed (approved):
One last fixup, but looks great to me overall. Thank you!

TamarZanzouri committed:
Update robot-server/robot_server/runs/run_store.py

Co-authored-by: Max Marrone <max@opentrons.com>

TamarZanzouri committed:
Merge branch 'edge' into EXEC-655-store-commands-error-list-in-db

TamarZanzouri committed:
import from schema_7
